### PR TITLE
Custom priorities and multiple renderers

### PR DESCRIPTION
**Pull request by colinodell**
Implements #339 and #35

See the CHANGELOG and UPGRADE files for more details

### CODE REVIEW

**Comment by Niklan** [comment]
First things first, thank you for this changes! It's even with patch is allows me to continue my work.

I do test with this changes for one of my custom syntax.

I've added my custom parser via new `register()` method for extensions.

`->addBlockParser(new MetaInformationParser(), 30)`

And this allows me to implement YAML-formatted metadata syntax support for my Markdown files!

```markdown
---
id: 1
title: Hello World
---
```

Before that, `ThematicBreakParser.php` take control over it and replaces `---` with `<hr />` every time, now I can control it! :confetti_ball: 

I'll do more testing, since I want to implement another different custom syntax which previously conflicted with default parsers. I write for any results later.

**Comment by Niklan** [comment]
Can't be sure if my current problem is related to this PR, but since I test it, so I ask here :)

I've trying to create Note markdown syntax [MDSN-like](https://docs.microsoft.com/en-us/contribute/how-to-write-use-markdown#note-blocks)

The syntax I test is

```markdown
> [!NOTE]
> This is a NOTE

> [!WARNING]
> This is a WARNING

> [!TIP]
> This is a TIP

> [!IMPORTANT]
> This is IMPORTANT

> Original Markdown Quote
```

I've created `NoteParser`

```php
<?php

namespace Drupal\druki_parser\CommonMark\Block\Parser;

use Drupal\druki_parser\CommonMark\Block\Element\NoteElement;
use League\CommonMark\Block\Parser\AbstractBlockParser;
use League\CommonMark\ContextInterface;
use League\CommonMark\Cursor;

/**
 * Class NoteParser
 *
 * @package Drupal\druki_parser\CommonMark\Block\Parser
 */
class NoteParser extends AbstractBlockParser {

  /**
   * {@inheritdoc}
   */
  public function parse(ContextInterface $context, Cursor $cursor): bool {
    if ($cursor->isIndented()) {
      return FALSE;
    }

    if ($cursor->getNextNonSpaceCharacter() !== '>') {
      return FALSE;
    }

    $cursor->advanceToNextNonSpaceOrTab();
    $cursor->advance();
    $cursor->advanceBySpaceOrTab();

    $matched = $cursor->match("/\[\!(NOTE|WARNING|TIP|IMPORTANT)\]/");
    if (!$matched) {
      return FALSE;
    }

    $context->addBlock(new NoteElement());

    return TRUE;
  }

}
```

`NoteElement`

```php
<?php

namespace Drupal\druki_parser\CommonMark\Block\Element;

use League\CommonMark\Block\Element\AbstractBlock;
use League\CommonMark\Cursor;

/**
 * Class NoteElement
 *
 * @package Drupal\druki_parser\Plugin\Markdown\Extension
 */
class NoteElement extends AbstractBlock {

  /**
   * {@inheritdoc}
   */
  public function canContain(AbstractBlock $block): bool {
    return FALSE;
  }

  /**
   * {@inheritdoc}
   */
  public function acceptsLines(): bool {
    return TRUE;
  }

  /**
   * {@inheritdoc}
   */
  public function isCode(): bool {
    return FALSE;
  }

  /**
   * {@inheritdoc}
   */
  public function shouldLastLineBeBlank(Cursor $cursor, $currentLineNumber): bool {
    return FALSE;
  }

  /**
   * {@inheritdoc}
   */
  public function matchesNextLine(Cursor $cursor): bool {
    dump('TEST');
    if (!$cursor->isIndented() && $cursor->getNextNonSpaceCharacter() === '>') {
      $cursor->advanceToNextNonSpaceOrTab();
      $cursor->advance();
      $cursor->advanceBySpaceOrTab();

      return TRUE;
    }

    return FALSE;
  }

}
```

And simple `NoteRenderer`

```php
<?php

namespace Drupal\druki_parser\CommonMark\Block\Renderer;

use League\CommonMark\Block\Element\AbstractBlock;
use League\CommonMark\Block\Renderer\BlockRendererInterface;
use League\CommonMark\ElementRendererInterface;
use League\CommonMark\HtmlElement;

/**
 * Class NoteRenderer
 *
 * @package Drupal\druki_parser\CommonMark\Renderer
 */
class NoteRenderer implements BlockRendererInterface {

  /**
   * {@inheritdoc}
   */
  public function render(AbstractBlock $block, ElementRendererInterface $htmlRenderer, $inTightList = FALSE): HtmlElement {
    $strings = $block->getStrings();
    dump($strings);

    return new HtmlElement('div', ['data-druki-note' => ''], []);
  }

}
```

Then I register it in my extension object.

```php
    $environment
      ->addBlockParser(new NoteParser(), 80)
      ->addBlockRenderer(
        'Drupal\druki_parser\CommonMark\Block\Element\NoteElement',
        new NoteRenderer()
      )
```

As you can see, I set priority to 80 (default `BlockQuoteParser` is 70), so my parser is override it.

And then replaced by renderer content.

The final result is not as I expected:

```html
<div data-druki-note=""></div>\n
<p>This is a NOTE</p>\n
<div data-druki-note=""></div>\n
<p>This is a WARNING</p>\n
<div data-druki-note=""></div>\n
<p>This is a TIP</p>\n
<div data-druki-note=""></div>\n
<p>This is IMPORTANT</p>\n
<p>Original Markdown Quote</p>\n
```

I didn't understand, but `NoteElement::matchesNextLine()` is never called, but other methods, like `canContain()` is called during processing.

This is related to this PR or I just miss something.

As I mention in previous comment, I created YAML-like meta syntax, it is also has `matchesNextLine()` and it called as expected.

**Comment by colinodell** [comment]
Try using xdebug to trace through `DocParser::incorporateLine()` to see why it might not be called.

**Comment by colinodell** [comment]
I was able to debug this today.  Basically, your `NoteElement` is saying that it cannot `contain` any child blocks.  You'll probably want to change that to something like this:

```php
public function canContain(AbstractBlock $block):
    return $block instanceof Paragraph;
}
```

You'll also need to modify your renderer to also render the child paragraph elements inside of your new `<div>`:

```php
public function render(AbstractBlock $block, ElementRendererInterface $htmlRenderer, $inTightList = FALSE): HtmlElement {
    $childContent = $htmlRenderer->renderBlocks($block->children());

    return new HtmlElement('div', ['data-druki-note' => ''], $childContent);
}
```

After making these changes I get the following output:

```html
<div data-druki-note=""><p>This is a NOTE</p></div>
<div data-druki-note=""><p>This is a WARNING</p></div>
<div data-druki-note=""><p>This is a TIP</p></div>
<div data-druki-note=""><p>This is IMPORTANT</p></div>
<p>Original Markdown Quote</p>
```

Notice how that original quote is no longer working?  That's because your custom parser is **always** moving the cursor past the `> ` **but it fails to restore the restore cursor when it realizes it does not match what you're looking for**.  You'll therefore want to modify your parser to look more like this:

```php
        $state = $cursor->saveState();

        $cursor->advanceToNextNonSpaceOrTab();
        $cursor->advance();
        $cursor->advanceBySpaceOrTab();

        $matched = $cursor->match("/\[\!(NOTE|WARNING|TIP|IMPORTANT)\]/");
        if (!$matched) {
            $cursor->restoreState($state);
            return FALSE;
        }
```

This will ensure the normal blockquote parser receives the cursor with it pointing to the `>` character and thus produce this output:

```html
<div data-druki-note=""><p>This is a NOTE</p></div>
<div data-druki-note=""><p>This is a WARNING</p></div>
<div data-druki-note=""><p>This is a TIP</p></div>
<div data-druki-note=""><p>This is IMPORTANT</p></div>
<blockquote>
<p>Original Markdown Quote</p>
</blockquote>
```

So overall it seems that my priority changes are working correctly, but I'd like you to complete your testing just to be sure.

Hope this feedback helps!

**Comment by Niklan** [comment]
Awesome, thanks! I did try later today this. Yesterday I can't handle it :man_shrugging: 

So, this definitely not related to PR, and I can handle rest of it by myself ;)

**Comment by Niklan** [comment]
Tested it on my project, works as expected! Thank you again.

At this moment looks like PR works great and implements my #339 issue.

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,31 @@ Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) princip
 
 ## [Unreleased][unreleased]
 
+### Added
+
+ - The priority of parsers, processors, and renderers can now be set when `add()`ing them; you no longer need to rely on the order in which they are added
+ - Added support for trying multiple parsers per block/inline
+ - Extracted two new bases interfaces from `Environment`:
```
Comment (glensc) [inline CHANGELOG.md]: `s/bases/base/`

Comment (colinodell) [reply CHANGELOG.md]: I'm slightly embarrassed by how many small issues like this are slipping through...  Thanks for catching this!

Comment (glensc) [reply CHANGELOG.md]: you may find https://grammarly.com/ browser extension to be useful making minor errors being hilited for you in text areas.

**Pull request merged**